File: src/main/java/priv/bajdcc/LALR1/grammar/tree/ExpCycleCtrl.java
package priv.bajdcc.LALR1.grammar.tree;

import priv.bajdcc.LALR1.grammar.codegen.ICodegen;
import priv.bajdcc.LALR1.grammar.semantic.ISemanticRecorder;
import priv.bajdcc.LALR1.grammar.tree.closure.IClosureScope;
import priv.bajdcc.util.lexer.token.KeywordType;
import priv.bajdcc.util.lexer.token.Token;

/**
 * 【语义分析】循环控制表达式
 *
 * @author bajdcc
 */
public class ExpCycleCtrl implements IExp {

	/**
	 * 变量名
	 */
	private Token name = null;

	public Token getName() {
		return name;
	}

	public void setName(Token name) {
		this.name = name;
	}

	@Override
	public boolean isConstant() {
		return false;
	}

	@Override
	public boolean isEnumerable() {
		return false;
	}

	@Override
	public IExp simplify(ISemanticRecorder recorder) {
		return this;
	}

	@Override
	public void analysis(ISemanticRecorder recorder) {

	}

	@Override
	public void genCode(ICodegen codegen) {
		KeywordType keyword = (KeywordType) name.object;
		if (keyword == KeywordType.BREAK) {
			codegen.getBlockService().genBreak();
		} else {
			codegen.getBlockService().genContinue();
		}
	}

	@Override
	public String toString() {
		return print(new StringBuilder());
	}

	@Override
	public String print(StringBuilder prefix) {
		return name.toRealString();
	}

	@Override
	public void addClosure(IClosureScope scope) {

	}

	@Override
	public void setYield() {

	}
}

File: src/main/java/com/bajdcc/LALR1/grammar/tree/ExpCycleCtrl.kt
package com.bajdcc.LALR1.grammar.tree

import com.bajdcc.LALR1.grammar.codegen.ICodegen
import com.bajdcc.LALR1.grammar.semantic.ISemanticRecorder
import com.bajdcc.LALR1.grammar.tree.closure.IClosureScope
import com.bajdcc.util.lexer.token.KeywordType
import com.bajdcc.util.lexer.token.Token

/**
 * 【语义分析】循环控制表达式
 *
 * @author bajdcc
 */
class ExpCycleCtrl : IExp {

    /**
     * 变量名
     */
    var name: Token? = null

    override fun isConstant(): Boolean {
        return false
    }

    override fun isEnumerable(): Boolean {
        return false
    }

    override fun simplify(recorder: ISemanticRecorder): IExp {
        return this
    }

    override fun analysis(recorder: ISemanticRecorder) {

    }

    override fun genCode(codegen: ICodegen) {
        val keyword = name!!.obj as KeywordType?
        if (keyword === KeywordType.BREAK) {
            codegen.blockService.genBreak()
        } else {
            codegen.blockService.genContinue()
        }
    }

    override fun toString(): String {
        return print(StringBuilder())
    }

    override fun print(prefix: StringBuilder): String {
        return name!!.toRealString()
    }

    override fun addClosure(scope: IClosureScope) {

    }

    override fun setYield() {

    }
}